stdlib/Test/src/GenericDimensionfuls.jl

```diff
@@ -0,0 +1,114 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+module GenericDimensionfuls
+
+export GenericDimensionful
+
+# Here we implement a minimal dimensionful type GenericDimensionful, which is used
+# to test dimensional correctness of various functions in Base.
+
+"""
+The `GenericDimensionful` type can be used to test that numerical code works with
+dimensionful quantities.   Construct a dimensionful quantity with `GenericDimensionful(x)`
+(or `GenericDimensionful{p}(x)` for unitsᵖ).  Arithmetic operations for `GenericDimensionful`
+are defined so that you can only combine quantities in ways that respect the units.
+"""
+struct GenericDimensionful{p,T<:Number} <: Number
+    val::T
+    function GenericDimensionful{p,T}(v::Number) where {p,T}
+        p isa Real || throw(DomainError("units must have real powers"))
+        new(v)
+    end
+end
+GenericDimensionful(x::T) where {T<:Number} = GenericDimensionful{1,T}(x)
+GenericDimensionful(x::GenericDimensionful) = x
+(::Type{T})(x::GenericDimensionful) where {T<:Number} = T(x.val)::T
```

Comment:
```suggestion
(::Type{T})(x::GenericDimensionful{0}) where {T<:Number} = T(x.val)::T
```
So that we don’t convert dimensionful numbers to dimensionless ones.